lib/Dialect/FIRRTL/Transforms/EmitMetadata.cpp

```diff
@@ -0,0 +1,249 @@
+  });
+  if (it != body->end()) {
+    auto instanceGraph = getAnalysis<InstanceGraph>();
+    auto *node = instanceGraph.lookup(&(*it));
+    llvm::for_each(llvm::depth_first(node), [&](InstanceGraphNode *node) {
+      dutModuleSet.insert(node->getModule());
+    });
+  }
+
+  // Any extmodule with these annotations or one of these ScalaClass classes
+  // should be excluded from the blackbox list.
+  auto *scalaClassAnnoClass = "sifive.enterprise.firrtl.ScalaClassAnnotation";
+  std::array<StringRef, 2> classBlackList = {
+      "freechips.rocketchip.util.BlackBoxedROM", "chisel3.shim.CloneModule"};
+  std::array<StringRef, 2> blackListedAnnos = {
+      "firrtl.transforms.BlackBoxInlineAnno",
+      "firrtl.transforms.BlackBoxResourceAnno"};
+
+  // Find all extmodules in the circuit. Check if they are black-listed from
+  // being included in the list. If they are not, separate them into two groups
+  // depending on if theyre in the DUT or the test harness.
+  SmallVector<StringRef> dutModules;
+  SmallVector<StringRef> testModules;
+  for (auto extModule : circuitOp.getBody()->getOps<FExtModuleOp>()) {
+    // If the module doesn't have a defname, then we can't record it properly.
+    // Just skip it.
+    if (!extModule.defname())
+      continue;
+
+    // If its a generated blackbox, skip it.
+    AnnotationSet annos(extModule);
+    if (llvm::any_of(blackListedAnnos, [&](auto blackListedAnno) {
+          return annos.hasAnnotation(blackListedAnno);
+        }))
+      continue;
+
+    // If its a blacklisted scala class, skip it.
+    auto scalaAnnoDict = annos.getAnnotation(scalaClassAnnoClass);
+    if (scalaAnnoDict) {
+      Annotation scalaAnno(scalaAnnoDict);
+      auto scalaClass = scalaAnno.getMember<StringAttr>("className");
+      if (scalaClass &&
+          llvm::is_contained(classBlackList, scalaClass.getValue()))
+        continue;
+    }
+
+    // Record the defname of the module.
+    if (dutModuleSet.contains(extModule)) {
+      dutModules.push_back(*extModule.defname());
+    } else {
+      testModules.push_back(*extModule.defname());
+    }
+  }
+
+  // This is a helper to create the verbatim output operation.
+  auto createOutput = [&](ArrayRef<StringRef> names, StringRef filename) {
+    if (filename.empty())
+      return;
+
+    // The output is a json array with each element a module name. The
+    // defname of a module can't change so we can output them verbatim.
+    std::string buffer;
+    llvm::raw_string_ostream os(buffer);
+    llvm::json::OStream j(os);
+    j.array([&] {
+      for (auto &name : names)
+        j.value(name);
+    });
+
+    // Put the information in a verbatim operation.
+    auto builder = OpBuilder::atBlockEnd(body);
+    auto verbatimOp =
+        builder.create<sv::VerbatimOp>(circuitOp.getLoc(), buffer);
+    auto fileAttr = hw::OutputFileAttr::getFromFilename(
+        context, filename, /*excludeFromFilelist=*/true);
+    verbatimOp->setAttr("output_file", fileAttr);
```

Comment:
Not needed here, but this `setAttr("output_file"` is begging for either a utility or a `getOutputFileKeyName()` to return the string.